crates/server/src/authority/catalog.rs

```diff
@@ -634,26 +638,99 @@ async fn send_authoritative_response(
+                        }
+                        Skip => {
+                            warn!("unexpected lookup skip");
+                            (None, None)
+                        }
+                    }
+                } else {
+                    (None, None)
+                }
+            }
+            #[cfg(not(feature = "dnssec"))]
             (None, None)
         }
     } else {
         let nsecs = if lookup_options.dnssec_ok() {
-            // in the dnssec case, nsec records should exist, we return NoError + NoData + NSec...
-            debug!("request: {} non-existent adding nsecs", request_id);
-            // run the nsec lookup future, and then transition to get soa
-            let future = authority.get_nsec_records(query.name(), lookup_options);
-            match future.await {
-                // run the soa lookup
-                Continue(Ok(nsecs)) | Break(Ok(nsecs)) => Some(nsecs),
-                Continue(Err(e)) | Break(Err(e)) => {
-                    warn!("failed to lookup nsecs: {}", e);
-                    None
-                }
-                Skip => {
-                    warn!("unexpected lookup skip");
-                    None
+            #[cfg(feature = "dnssec")]
+            {
+                // in the dnssec case, nsec records should exist, we return NoError + NoData + NSec...
+                debug!("request: {_request_id} non-existent adding nsecs");
+                match authority.nx_proof_kind() {
+                    Some(nx_proof_kind) => {
+                        // run the nsec lookup future, and then transition to get soa
+                        let future = match nx_proof_kind {
+                            NxProofKind::Nsec => {
+                                authority.get_nsec_records(query.name(), lookup_options)
+                            }
+                            NxProofKind::Nsec3 {
+                                algorithm,
+                                salt,
+                                iterations,
+                            } => authority.get_nsec3_records(
+                                Nsec3QueryInfo {
+                                    qname: query.name(),
+                                    qtype: query.query_type(),
+                                    has_wildcard_match: false,
+                                    algorithm: *algorithm,
+                                    salt,
+                                    iterations: *iterations,
+                                },
+                                lookup_options,
+                            ),
+                        };
+
+                        match future.await {
+                            // run the soa lookup
+                            Continue(Ok(nsecs)) | Break(Ok(nsecs)) => Some(nsecs),
+                            Continue(Err(e)) | Break(Err(e)) => {
+                                warn!("failed to lookup nsecs: {e}");
```

Comment:
can we add the request_id to this? seems like that would be more useful on the warning.